cpp/src/arrow/array/concatenate.cc

```diff
@@ -436,6 +438,28 @@ class ConcatenateImpl {
     return Status::OK();
   }
 
+  Status Visit(const RunEndEncodedType& type) {
+    int64_t physical_length = 0;
+    for (const auto& input : in_) {
+      if (internal::AddWithOverflow(physical_length,
+                                    ree_util::FindPhysicalLength(ArraySpan(*input)),
+                                    &physical_length) ||
+          // Make sure we can safely downcast to int32_t
```

Comment:
Hmm, why do we have to downcast to int32_t here? Shouldn't it be based on the offsets type?

Reply:
This is from @zagto and I assumed it was due to https://arrow.apache.org/docs/dev/format/Columnar.html#array-lengths

But note that run-ends are cumulative sums of the run-lengths, so summing the lengths of `INT_MAX` runs wouldn't really fit in 32-bit integers.

Reply:
Hmm, I don't think we make this check for any other types though (if anything, I think we have tests that check very long arrays)

Reply:
The implementation of `AppendArraySlice` should check for all of that.

Can I just remove the physical length sum check, then?

Reply:
Ok, if there's already a check, maybe note that this one is redundant and remove it (is there a test covering that though?)

Reply:
Done.